tee/prototype/epochproc.go

```diff
@@ -0,0 +1,397 @@
+		proofs[i] = bp
+	}
+	return proofs
+}
+
+// prune prunes a given slice of transactions for a given set of addresses, s.t.
+// no element of the given set is either a recipient or sender in a transaction,
+// it also makes sure to remove exited parties from the txEpoch and the
+// balance proof cache.
+func prune(epoch *Epoch,
+	txs txCache,
+	bpc balProofCache,
+	forS []common.Address,
+) (*Epoch, balProofCache, []*tee.Transaction) {
+	for _, m := range forS {
+		delete(epoch.balances, m) // remove member from txEpoch
+		delete(bpc, m)            // remove member from balProofCache
+
+		sTxs, sOk := txs.senders[m]
+		rTxs, rOk := txs.recipients[m]
+		if !sOk && !rOk {
+			continue
+		}
+		for _, tx := range sTxs {
+			delete(txs.txs, tx.Hash())
+		}
+		for _, tx := range rTxs {
+			delete(txs.txs, tx.Hash())
+		}
+	}
+	prunedTxs := make([]*tee.Transaction, 0, len(txs.txs))
+	i := 0
+	for _, tx := range txs.txs {
+		prunedTxs[i] = tx
+	}
+	return epoch, bpc, prunedTxs
+}
+
+type balProofCache = map[common.Address]*tee.BalanceProof
+
+func (e *Enclave) depositExitRoutine(
+	verifiedBlocks <-chan *tee.Block,
+	phaseDone chan<- exitersSet,
+	depositEpoch, exitEpoch *Epoch,
+) error {
+	// read blocks from verifiedBlocks (deposit phase).
+	for vb := range verifiedBlocks {
+		exiters, err := e.handleVerifiedBlock(depositEpoch, exitEpoch, vb)
+		if err != nil {
+			return fmt.Errorf("handling verified blocknr %v : %w", vb.NumberU64(), err)
+		}
+		if e.phaseDone(vb.NumberU64()) {
+			e.depositProofs <- retrieveCachedDepProofs(e.depositProofCache)
+			e.depositProofCache = make(map[common.Address]*tee.DepositProof)
+			phaseDone <- exiters
+			close(phaseDone) // phase done, stop TX processor.
+			return nil
+		}
+	}
+	return nil
+}
+
+// retrieveCachedDepProofs reduces the deposit proof cache to a slice of `tee.DepositProof`s.
+func retrieveCachedDepProofs(cache map[common.Address]*tee.DepositProof) []*tee.DepositProof {
+	dps := make([]*tee.DepositProof, 0, len(cache))
+	for _, dp := range cache {
+		dps = append(dps, dp)
+	}
+	return dps
+}
+
+// generateDepositProof generates the deposit proof for the given user in the given
+// deposit Epoch.
+func (e *Enclave) generateDepositProof(depEpoch *Epoch, acc common.Address) (*tee.DepositProof, error) {
+	b := tee.Balance{
+		Epoch:   depEpoch.Number,
+		Account: acc,
+		Value:   new(big.Int).Set(depEpoch.balances[acc].Value),
+	}
+
+	msg, err := tee.EncodeDepositProof(e.params.Contract, b)
+	if err != nil {
+		return nil, fmt.Errorf("encoding deposit proof: %w", err)
+	}
+
+	sig, err := e.wallet.SignText(e.account, crypto.Keccak256(msg))
+	if err != nil {
+		return nil, fmt.Errorf("signing deposit proof: %w", err)
+	}
+
+	return &tee.DepositProof{
+		Balance: b,
+		Sig:     sig,
+	}, nil
+}
+
+// updateBalanceProofs updates the balance proof cache with for given transaction.
+func (e *Enclave) updateBalanceProofs(bpc balProofCache, ep *Epoch, sender, recipient common.Address) (balProofCache, error) {
```

Comment:
a map is a pointer type, no need to return it, you work on the same instance as the caller.

Reply:
This is true. The reason I went for this is, because in the usage it's clear that `balProofCache` is being modified within that function. Feels more functional :see_no_evil:  So looking at

``` Go
		bpc, err = e.updateBalanceProofs(bpc, ep, tx.Sender, tx.Recipient)
		if err != nil {
			return nil, fmt.Errorf("applying tx in txEpoch %v: %w", ep.Number, err)
		}
```
makes it obvious that bpc gets changed. What do you think?

Reply:
If we directly apply, instead of cache, we don't use this anymore, right?

Reply:
Got rid of it. :ok_hand: